File: templates/problems/11_graphs/12_redundant_connection/go/redundant_connection_test.go
package redundant_connection

import (
	"reflect"
	"testing"
)

func TestFindRedundantConnection(t *testing.T) {
	tests := []struct {
		name  string
		edges [][]int
		want  []int
	}{
		{
			name:  "triangle",
			edges: [][]int{{1, 2}, {1, 3}, {2, 3}},
			want:  []int{2, 3},
		},
		{
			name:  "four nodes with cycle",
			edges: [][]int{{1, 2}, {2, 3}, {3, 4}, {1, 4}, {1, 5}},
			want:  []int{1, 4},
		},
		{
			name:  "last edge creates cycle",
			edges: [][]int{{1, 2}, {1, 3}, {3, 4}, {2, 4}},
			want:  []int{2, 4},
		},
		{
			name:  "two nodes",
			edges: [][]int{{1, 2}, {2, 1}},
			want:  []int{2, 1},
		},
		{
			name:  "five node cycle",
			edges: [][]int{{1, 2}, {2, 3}, {3, 4}, {4, 5}, {5, 1}},
			want:  []int{5, 1},
		},
		{
			name:  "star with extra edge",
			edges: [][]int{{1, 2}, {1, 3}, {1, 4}, {3, 4}},
			want:  []int{3, 4},
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got := findRedundantConnection(tt.edges)
			if !reflect.DeepEqual(got, tt.want) {
				t.Errorf("findRedundantConnection(%v) = %v, want %v", tt.edges, got, tt.want)
			}
		})
	}
}

File: templates/problems/11_graphs/12_redundant_connection/typescript/redundant_connection.test.ts
import { describe, it, expect } from "vitest";
import { findRedundantConnection } from "./redundant_connection";

describe("findRedundantConnection", () => {
    it.each([
        { name: "triangle", edges: [[1, 2], [1, 3], [2, 3]], want: [2, 3] },
        { name: "four nodes with cycle", edges: [[1, 2], [2, 3], [3, 4], [1, 4], [1, 5]], want: [1, 4] },
        { name: "last edge creates cycle", edges: [[1, 2], [1, 3], [3, 4], [2, 4]], want: [2, 4] },
        { name: "two nodes", edges: [[1, 2], [2, 1]], want: [2, 1] },
        { name: "five node cycle", edges: [[1, 2], [2, 3], [3, 4], [4, 5], [5, 1]], want: [5, 1] },
        { name: "star with extra edge", edges: [[1, 2], [1, 3], [1, 4], [3, 4]], want: [3, 4] },
    ])("case: $name", ({ edges, want }) => {
        const got = findRedundantConnection(edges);
        expect(got).toEqual(want);
    });
});
